src/ray/rpc/retryable_grpc_client.h

```diff
@@ -0,0 +1,306 @@
+    std::weak_ptr<RetryableGrpcClient> weak_self = shared_from_this();
+    timer_->async_wait([weak_self](boost::system::error_code error) {
+      if (auto self = weak_self.lock()) {
+        if (error == boost::system::errc::success) {
+          self->CheckChannelStatus();
+        }
+      }
+    });
+  }
+
+  void CheckChannelStatus(bool reset_timer = true) {
+    if (shutdown_) {
+      return;
+    }
+
+    auto status = channel_->GetState(false);
+    // https://grpc.github.io/grpc/core/md_doc_connectivity-semantics-and-api.html
+    // https://grpc.github.io/grpc/core/connectivity__state_8h_source.html
+    if (status != GRPC_CHANNEL_READY) {
+      RAY_LOG(DEBUG) << "GRPC channel status: " << status;
+    }
+
+    // We need to cleanup all the pending requests which are timeout.
+    auto now = absl::Now();
+    while (!pending_requests_.empty()) {
+      auto iter = pending_requests_.begin();
+      if (iter->first > now) {
+        break;
+      }
+      auto [executor, request_bytes] = iter->second;
+      executor->Abort(ray::Status::TimedOut(absl::StrFormat(
+          "Timed out while waiting for %s to become available.", server_name_)));
+      pending_requests_bytes_ -= request_bytes;
+      delete executor;
+      pending_requests_.erase(iter);
+    }
+
+    switch (status) {
+    case GRPC_CHANNEL_TRANSIENT_FAILURE:
+    case GRPC_CHANNEL_CONNECTING:
+      if (!server_is_unavailable_) {
+        server_is_unavailable_ = true;
+      } else {
+        uint64_t server_unavailable_duration_seconds = static_cast<uint64_t>(
+            absl::ToInt64Seconds(absl::Now() - server_last_available_time_));
+        if (server_unavailable_duration_seconds >= server_unavailable_timeout_seconds_) {
+          RAY_LOG(WARNING) << server_name_ << " has been unavailable for "
+                           << server_unavailable_duration_seconds << " seconds";
+          server_unavailable_timeout_callback_();
+        }
+      }
+      break;
+    case GRPC_CHANNEL_SHUTDOWN:
+      RAY_CHECK(shutdown_) << "Channel shoud never go to this status.";
+      break;
+    case GRPC_CHANNEL_READY:
+    case GRPC_CHANNEL_IDLE:
+      server_last_available_time_ = absl::Now();
+      server_is_unavailable_ = false;
+      // Retry the ones queued.
+      while (!pending_requests_.empty()) {
+        pending_requests_.begin()->second.first->Retry();
+        pending_requests_.erase(pending_requests_.begin());
+      }
+      pending_requests_bytes_ = 0;
+      break;
+    default:
+      RAY_LOG(FATAL) << "Not covered status: " << status;
+    }
+
+    SetupCheckTimer();
+  }
+
+  instrumented_io_context &io_context_;
+  // Timer can be called from either the GRPC event loop, or the application's
+  // main thread. It needs to be protected by a mutex.
+  absl::Mutex timer_mu_;
+  const std::unique_ptr<boost::asio::deadline_timer> timer_;
+
+  std::shared_ptr<grpc::Channel> channel_;
+
+  uint64_t max_pending_requests_bytes_;
```

Comment:
const for these params, to make them distinguishable from mutable ones like server_is_unavailable_.